{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2b Mash filtration & clustering\n",
    "\n",
    "In this notebook, we run Mash to generate genome-wise pairwise similarity scores (which correspond to Average Nucleotide Identify (ANI) & DNA-DNA reassociation value).\n",
    "\n",
    "Mash will be used as a final filtration metric to filter out strains which are too dissimilar from the rest of the genome collection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.cluster.hierarchy as hc\n",
    "import scipy.spatial as sp\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.patches as patches\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from kneebow.rotor import Rotor\n",
    "\n",
    "# pyphyon import\n",
    "import pyphylon.mash as mash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW = \"data/raw/\"\n",
    "FNA_GENOMES = \"data/raw/genomes/fna/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(241, 20)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>genome_status</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>contig_l50</th>\n",
       "      <th>contig_n50</th>\n",
       "      <th>chromosomes</th>\n",
       "      <th>plasmids</th>\n",
       "      <th>contigs</th>\n",
       "      <th>patric_cds</th>\n",
       "      <th>refseq_cds</th>\n",
       "      <th>trna</th>\n",
       "      <th>rrnacoarse_consistency</th>\n",
       "      <th>fine_consistency</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>genome_qualitydate_created</th>\n",
       "      <th>date_modified</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1010840.4</td>\n",
       "      <td>Streptococcus pyogenes MGAS1882</td>\n",
       "      <td>1010840</td>\n",
       "      <td>Complete</td>\n",
       "      <td>1781029</td>\n",
       "      <td>38.00</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1727</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015-03-16T03:17:09.594Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1048264.3</td>\n",
       "      <td>Streptococcus pyogenes HKU QMH11M0907901</td>\n",
       "      <td>1048264</td>\n",
       "      <td>Complete</td>\n",
       "      <td>1908100</td>\n",
       "      <td>38.45</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1909</td>\n",
       "      <td>1865</td>\n",
       "      <td>67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-01-17T15:29:01.552Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1150773.3</td>\n",
       "      <td>Streptococcus pyogenes JRS4</td>\n",
       "      <td>1150773</td>\n",
       "      <td>Complete</td>\n",
       "      <td>1811968</td>\n",
       "      <td>38.63</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1811</td>\n",
       "      <td>1671</td>\n",
       "      <td>67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.8</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-01-17T16:03:54.402Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1150773.4</td>\n",
       "      <td>Streptococcus pyogenes JRS4</td>\n",
       "      <td>1150773</td>\n",
       "      <td>Complete</td>\n",
       "      <td>1811124</td>\n",
       "      <td>38.64</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1886</td>\n",
       "      <td>1890</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2016-03-01T06:31:23.641Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1235829.3</td>\n",
       "      <td>Streptococcus pyogenes A20</td>\n",
       "      <td>1235829</td>\n",
       "      <td>Complete</td>\n",
       "      <td>1837281</td>\n",
       "      <td>38.54</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1834</td>\n",
       "      <td>1828</td>\n",
       "      <td>67</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.4</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2015-03-16T03:17:09.594Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genome_id                               genome_name taxon_id genome_status  \\\n",
       "0  1010840.4           Streptococcus pyogenes MGAS1882  1010840      Complete   \n",
       "1  1048264.3  Streptococcus pyogenes HKU QMH11M0907901  1048264      Complete   \n",
       "2  1150773.3               Streptococcus pyogenes JRS4  1150773      Complete   \n",
       "3  1150773.4               Streptococcus pyogenes JRS4  1150773      Complete   \n",
       "5  1235829.3                Streptococcus pyogenes A20  1235829      Complete   \n",
       "\n",
       "   genome_length  gc_content  contig_l50  contig_n50 chromosomes plasmids  \\\n",
       "0        1781029       38.00           1           1           1        0   \n",
       "1        1908100       38.45           1           1           1      NaN   \n",
       "2        1811968       38.63           1           1           1      NaN   \n",
       "3        1811124       38.64           1           1           1      NaN   \n",
       "5        1837281       38.54           1           1           1        0   \n",
       "\n",
       "   contigs  patric_cds refseq_cds trna rrnacoarse_consistency  \\\n",
       "0        1        1727          0   57                    NaN   \n",
       "1        1        1909       1865   67                    NaN   \n",
       "2        1        1811       1671   67                    NaN   \n",
       "3        1        1886       1890   66                    NaN   \n",
       "5        1        1834       1828   67                    NaN   \n",
       "\n",
       "  fine_consistency  checkm_completeness  checkm_contamination  \\\n",
       "0             99.5                100.0                   0.0   \n",
       "1             99.9                100.0                   0.9   \n",
       "2             99.8                100.0                   0.0   \n",
       "3             99.9                100.0                   0.0   \n",
       "5             99.4                100.0                   0.0   \n",
       "\n",
       "  genome_qualitydate_created             date_modified  \n",
       "0                        NaN  2015-03-16T03:17:09.594Z  \n",
       "1                        NaN  2016-01-17T15:29:01.552Z  \n",
       "2                        NaN  2016-01-17T16:03:54.402Z  \n",
       "3                        NaN  2016-03-01T06:31:23.641Z  \n",
       "5                        NaN  2015-03-16T03:17:09.594Z  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(241, 66)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>genome_name</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>genome_status</th>\n",
       "      <th>strain</th>\n",
       "      <th>serovar</th>\n",
       "      <th>biovar</th>\n",
       "      <th>pathovar</th>\n",
       "      <th>mlst</th>\n",
       "      <th>...</th>\n",
       "      <th>motility</th>\n",
       "      <th>sporulation</th>\n",
       "      <th>temperature_range</th>\n",
       "      <th>optimal_temperature</th>\n",
       "      <th>salinity</th>\n",
       "      <th>oxygen_requirement</th>\n",
       "      <th>habitat</th>\n",
       "      <th>disease</th>\n",
       "      <th>comments</th>\n",
       "      <th>additional_metadata</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1010840.4</td>\n",
       "      <td>Streptococcus pyogenes MGAS1882</td>\n",
       "      <td>Streptococcus pyogenes MGAS1882</td>\n",
       "      <td>1010840</td>\n",
       "      <td>Complete</td>\n",
       "      <td>MGAS1882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLST.Streptococcus_pyogenes.172</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Host</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1048264.3</td>\n",
       "      <td>Streptococcus pyogenes HKU QMH11M0907901</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1048264</td>\n",
       "      <td>Complete</td>\n",
       "      <td>HKU QMH11M0907901</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLST.Streptococcus_pyogenes.36</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Clinical use of next generation sequencing for...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1150773.3</td>\n",
       "      <td>Streptococcus pyogenes JRS4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1150773</td>\n",
       "      <td>Complete</td>\n",
       "      <td>JRS4</td>\n",
       "      <td>serovar emm6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLST.Streptococcus_pyogenes.37</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>We report the complete genome assemblies of th...</td>\n",
       "      <td>collected_by:Rockefeller University Lancefield...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1150773.4</td>\n",
       "      <td>Streptococcus pyogenes JRS4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1150773</td>\n",
       "      <td>Complete</td>\n",
       "      <td>JRS4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLST.Streptococcus_pyogenes.37</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pharyngitis</td>\n",
       "      <td>Complete genome sequence of the highly invasiv...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1235829.3</td>\n",
       "      <td>Streptococcus pyogenes A20</td>\n",
       "      <td>Streptococcus pyogenes A20</td>\n",
       "      <td>1235829</td>\n",
       "      <td>Complete</td>\n",
       "      <td>A20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MLST.Streptococcus_pyogenes.28</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Necrotizing faciitis</td>\n",
       "      <td>A clinical strain was isolated from a Necrotiz...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 66 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   genome_id                               genome_name  \\\n",
       "0  1010840.4           Streptococcus pyogenes MGAS1882   \n",
       "1  1048264.3  Streptococcus pyogenes HKU QMH11M0907901   \n",
       "2  1150773.3               Streptococcus pyogenes JRS4   \n",
       "3  1150773.4               Streptococcus pyogenes JRS4   \n",
       "5  1235829.3                Streptococcus pyogenes A20   \n",
       "\n",
       "                     organism_name taxon_id genome_status             strain  \\\n",
       "0  Streptococcus pyogenes MGAS1882  1010840      Complete           MGAS1882   \n",
       "1                              NaN  1048264      Complete  HKU QMH11M0907901   \n",
       "2                              NaN  1150773      Complete               JRS4   \n",
       "3                              NaN  1150773      Complete               JRS4   \n",
       "5       Streptococcus pyogenes A20  1235829      Complete                A20   \n",
       "\n",
       "        serovar biovar pathovar                             mlst  ...  \\\n",
       "0           NaN    NaN      NaN  MLST.Streptococcus_pyogenes.172  ...   \n",
       "1           NaN    NaN      NaN   MLST.Streptococcus_pyogenes.36  ...   \n",
       "2  serovar emm6    NaN      NaN   MLST.Streptococcus_pyogenes.37  ...   \n",
       "3           NaN    NaN      NaN   MLST.Streptococcus_pyogenes.37  ...   \n",
       "5           NaN    NaN      NaN   MLST.Streptococcus_pyogenes.28  ...   \n",
       "\n",
       "  motility sporulation temperature_range optimal_temperature salinity  \\\n",
       "0      NaN         NaN               NaN                 NaN      NaN   \n",
       "1      NaN         NaN               NaN                 NaN      NaN   \n",
       "2      NaN         NaN               NaN                   C      NaN   \n",
       "3       No         NaN               NaN                 NaN      NaN   \n",
       "5      NaN         NaN               NaN                 NaN      NaN   \n",
       "\n",
       "  oxygen_requirement habitat               disease  \\\n",
       "0                NaN    Host                   NaN   \n",
       "1                NaN     NaN                   NaN   \n",
       "2                NaN     NaN                   NaN   \n",
       "3                NaN     NaN           Pharyngitis   \n",
       "5                NaN     NaN  Necrotizing faciitis   \n",
       "\n",
       "                                            comments  \\\n",
       "0                                                  -   \n",
       "1  Clinical use of next generation sequencing for...   \n",
       "2  We report the complete genome assemblies of th...   \n",
       "3  Complete genome sequence of the highly invasiv...   \n",
       "5  A clinical strain was isolated from a Necrotiz...   \n",
       "\n",
       "                                 additional_metadata  \n",
       "0                                                NaN  \n",
       "1                                                NaN  \n",
       "2  collected_by:Rockefeller University Lancefield...  \n",
       "3                                                NaN  \n",
       "5                                                NaN  \n",
       "\n",
       "[5 rows x 66 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "scrubbed_species_summary = pd.read_pickle('data/interim/genome_summary_2a.pickle')\n",
    "scrubbed_species_metadata = pd.read_pickle('data/interim/genome_metadata_2a.pickle')\n",
    "\n",
    "\n",
    "display(\n",
    "    scrubbed_species_summary.shape,\n",
    "    scrubbed_species_summary.head(),\n",
    "    scrubbed_species_metadata.shape,\n",
    "    scrubbed_species_metadata.head()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run Mash\n",
    "\n",
    "- Running Mash on terminal:\n",
    "    - `mash sketch -o data/raw/combined_sketch.msh data/raw/genomes/fna/*.fna`\n",
    "- Running Mash using pyphylon:\n",
    "    - Run the following cell below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "CalledProcessError",
     "evalue": "Command '['mash', 'sketch', '-o', 'data/raw/combined_sketch.msh', 'data/raw/genomes/fna/*.fna']' returned non-zero exit status 127.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[4], line 8\u001b[0m\n\u001b[1;32m      1\u001b[0m cmd \u001b[38;5;241m=\u001b[39m [\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmash\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msketch\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m-o\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata/raw/combined_sketch.msh\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdata/raw/genomes/fna/*.fna\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      5\u001b[0m ]\n\u001b[0;32m----> 8\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mshell\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstdout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPIPE\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstderr\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPIPE\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# Print the output and error (if any)\u001b[39;00m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSTDOUT:\u001b[39m\u001b[38;5;124m\"\u001b[39m, result\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mdecode())\n",
      "File \u001b[0;32m~/miniconda3/envs/pyphylon-test/lib/python3.11/subprocess.py:571\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    569\u001b[0m     retcode \u001b[38;5;241m=\u001b[39m process\u001b[38;5;241m.\u001b[39mpoll()\n\u001b[1;32m    570\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m check \u001b[38;5;129;01mand\u001b[39;00m retcode:\n\u001b[0;32m--> 571\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m CalledProcessError(retcode, process\u001b[38;5;241m.\u001b[39margs,\n\u001b[1;32m    572\u001b[0m                                  output\u001b[38;5;241m=\u001b[39mstdout, stderr\u001b[38;5;241m=\u001b[39mstderr)\n\u001b[1;32m    573\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m CompletedProcess(process\u001b[38;5;241m.\u001b[39margs, retcode, stdout, stderr)\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['mash', 'sketch', '-o', 'data/raw/combined_sketch.msh', 'data/raw/genomes/fna/*.fna']' returned non-zero exit status 127."
     ]
    }
   ],
   "source": [
    "result = mash.sketch_genomes(genome_dir='data/raw/genomes/fna/*.fna', output_file='data/raw/combined_sketch.msh')\n",
    "\n",
    "# Print the output and error (if any)\n",
    "print(\"STDOUT:\", result.stdout.decode())\n",
    "print(\"STDERR:\", result.stderr.decode())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyphylon-test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
